### DISCUSSION TITLE
Tracing on handlers

### DISCUSSION THREAD

**Discussion by elbaro** [rwf2/Rocket#2053] [Questions] [answered]
Until Rocket has a built-in support for tracing, I am wrapping my handler with `tracing::instrument` and it works fine.

```rs
#[get("/abc?<params..>")]
pub async fn abc() -> Result<rocket::response::content::Json<String>, rocket::http::Status> {
  let task = async {
    ..
  };
  task.instrument(tracing::info_span!(
        "GET /abc"
    )).await
}
```

Now I want to add a `http.status` to the span.

Method1. Fairing + local_cache

```rs
#[rocket::async_trait]
impl Fairing for TracingFairing {
    fn info(&self) -> Info {
        todo!()
    }
    async fn on_request(&self, req: &mut Request<'_>, _data: &mut Data<'_>) {
        req.local_cache(|| {
            info_span!(
                &format!("{} {}", req.method(), req.url().path()),
            )
        });
    }

    async fn on_response<'r>(&self, req: &'r Request<'_>, res: &mut Response<'r>) {
        let span: &Option<&Span> = req.local_cache(|| None);
        if let Some(span) = span {
            span.record("http.status", &res.status().code);
        }
    }
}
```

it turns out `info_span!` only supports the static string, so `Fairing::on_request` cannot be used.

Method2.
Creates a local cache in the handler. However the handler does not accept &Request as a parameter:

```rs
#[get("/abc?<params..>")]
pub async fn abc(req: &Request<'_>) -> Result<rocket::response::content::Json<String>, rocket::http::Status> {
    req.local_cache(|| {
        info_span!("/abc")
    });
  ..
}
```

Method3.
Without Fairing+local_cache, the handler tries to get the status code.
However using `Responder::respond_to(&req)` requires the request, the same problem as the above.

```rs
#[get("/abc?<params..>")]
pub async fn abc(req: &Request<'_>) -> Result<rocket::response::content::Json<String>, rocket::http::Status> {
  let task = async { .. };
  let result = task.await.respond_to(&req);
  let code = match result {
    Ok(response) => response.status().code,
    Err(status) => status.code,
  };
  span.record("http.status", code);
  response
}
```

Is &Request as a parameter prohibited by design? Is there a good way to get status code, or add local_cache in the handler?

**Comment by somehowchris** [comment, marked as answer]
* `Method #1`
 Do you have an example, I'm sure this would work
* `Method #2 & #3`
 Why would you need any of the request? You know your route as you specified it and you know what you gonna give back so you know what status code it will by, no need to call something with `respond_to` or what ever. Just do you business logic and you will know what your outcome is.

In general, please give more concrete examples, there are variables used which aren't declared, no clue what `task` even is and what it returns that you could call a function on it.

**Reply by elbaro** [reply]
Method1

You can try compiling this module:

```rs
use rocket::{
    fairing::{Fairing, Info},
    Data, Request, Response,
};
use tracing::{info_span, Span};

struct TracingFairing {}

#[rocket::async_trait]
impl Fairing for TracingFairing {
    fn info(&self) -> Info {
        todo!()
    }
    async fn on_request(&self, req: &mut Request<'_>, _data: &mut Data<'_>) {
        req.local_cache(|| info_span!(&format!("{} {}", req.method(), req.url().path()),));
    }

    async fn on_response<'r>(&self, req: &'r Request<'_>, res: &mut Response<'r>) {
        let span: &Option<&Span> = req.local_cache(|| None);
        if let Some(span) = span {
            span.record("http.status", &res.status().code);
        }
    }
}
```

```
   |
15 |         req.local_cache(|| info_span!(&format!("{} {}", req.method(), req.url().path()),));
   |                            -----------------------------^^^------------------------------
   |                            |                            |
   |                            |                            non-constant value
   |                            help: consider using `let` instead of `static`: `let CALLSITE`

error[E0435]: attempt to use a non-constant value in a constant
   |
15 |         req.local_cache(|| info_span!(&format!("{} {}", req.method(), req.url().path()),));
   |                            -------------------------------------------^^^----------------
   |                            |                                          |
   |                            |                                          non-constant value
   |                            help: consider using `let` instead of `static`: `let CALLSITE`
```
https://github.com/tokio-rs/tracing/issues/1314
https://github.com/tokio-rs/tracing/pull/1020

Method2
Can I set local_cache without the request object?

Method3
No, I know the responder object, not the status code. Yes, it can be inferred.

Can you guess the status code for this `impl Responder` at a glance:
```rs
let ret: Result<
    rocket::response::content::Json<
        Result<serde_json::Value, rocket::response::status::BadRequest<serde_json::Value>>,
    >,
    Status,
  > =
  Ok(rocket::response::content::Json(Err(
      rocket::response::status::BadRequest(Some(serde_json::json!({ "error": err_msg }))),
  )));
```

I have many handlers with complex `impl Responder`s, inferring and adding the code to each is hard and vulnerable to mistakes.
For example, since `BadRequest` already implies the code 400, I don't want to write `return (BadRequest, 400)`.

**Reply by somehowchris** [reply]
Might this example help? https://github.com/somehowchris/rocket-tracing-fairing-example

**Reply by elbaro** [reply]
Thanks for the example.

I cloned the above repo and changed the tracing span name "request" to dynamic one, and it errors the same.

```rs
let span = info_span!(
    // "request",
    &format!("{} {}", req.method(), req.uri().path()),
    method = %req.method(),
    uri = %req.uri().path(),
    user_agent,
    status_code = tracing::field::Empty,
    request_id=%request_id.0
);
```

`method` and `uri` should be included in the span name because some tracing backends (like datadog) identifies the spans by its name only. With "request" as a name, spans from different url/handlers are not distinguishable.

![image](https://user-images.githubusercontent.com/1851290/148611081-64646a8e-5506-4dc9-a3f6-532c88377393.png)

^ When clicking a single span it shows its attributes (method and uri). However one cannot filter spans by method or uri.

**Reply by somehowchris** [reply]
Well that seems to be a limitation of the tracing crate, as the name needs to be constant and not dynamic

**Reply by somehowchris** [reply]
Might this be the solution? https://github.com/tokio-rs/tracing/pull/732

**Reply by somehowchris** [reply]
And again, as said, if you don‘t have an example with data dog, it‘s hard to provide options

**Reply by elbaro** [reply]
`otel.name` worked great.

```rs
let user_agent = req.headers().get_one("User-Agent").unwrap_or("");
let span_name = format!("{} {}", req.method(), req.uri().path());

let span = info_span!(
    "",
    otel.name = %span_name,
    http.method = %req.method(),
    http.url = %req.uri().path(),
    http.useragent = user_agent,
    http.status_code = tracing::field::Empty,
);
```

**Reply by somehowchris** [reply]
Hope to get this upstream without all of that boilerplate as soon as Sergio will return
